packages/gatsby/src/services/start-webpack-server.ts

```diff
@@ -0,0 +1,117 @@
+import openurl from "better-opn"
+import report from "gatsby-cli/lib/reporter"
+import formatWebpackMessages from "react-dev-utils/formatWebpackMessages"
+import chalk from "chalk"
+import { Compiler } from "webpack"
+
+import {
+  reportWebpackWarnings,
+  structureWebpackErrors,
+} from "../utils/webpack-error-utils"
+
+import { printDeprecationWarnings } from "../utils/print-deprecation-warnings"
+import { printInstructions } from "../utils/print-instructions"
+import { prepareUrls } from "../utils/prepare-urls"
+import { startServer } from "../utils/start-server"
+import { WebsocketManager } from "../utils/websocket-manager"
+import { IBuildContext } from "./"
+import {
+  markWebpackStatusAsPending,
+  markWebpackStatusAsDone,
+} from "../utils/webpack-status"
+import { enqueueFlush } from "../utils/page-data"
+
+export async function startWebpackServer({
+  program,
+  app,
+  workerPool,
+}: Partial<IBuildContext>): Promise<{
+  compiler: Compiler
+  websocketManager: WebsocketManager
+}> {
+  if (!program || !app) {
+    throw new Error(`Missing required params`)
+  }
+  let { compiler, webpackActivity, websocketManager } = await startServer(
+    program,
+    app,
+    workerPool
+  )
+
+  compiler.hooks.invalid.tap(`log compiling`, function () {
+    markWebpackStatusAsPending()
+  })
+
+  compiler.hooks.watchRun.tapAsync(`log compiling`, function (_, done) {
+    if (webpackActivity) {
+      webpackActivity.end()
+    }
+    webpackActivity = report.activityTimer(`Re-building development bundle`, {
+      id: `webpack-develop`,
+    })
+    webpackActivity.start()
+
+    done()
+  })
+
+  let isFirstCompile = true
+
+  return new Promise(resolve => {
```

Comment:
What's the use case of this `Promise` over returning `{ compiler, websocketManager }` below, rather than calling `resolve` explicitly? The function is already `async`.

Reply:
The reason is because we're having to promisify the webpack `done` hook. If we return it right away it won't have finished launching. By returning a promise that resolves when it's done we can hide that implementaiton detail and treat it as a normal async function.

Reply:
Ah, apparently `compiler.hooks.done.tapAsync` does not return a promise. A bit confusing but definitely a good reason to require promise wrapping.